xarray/core/combine.py

```diff
@@ -369,34 +370,178 @@ def _auto_concat(datasets, dim=None, data_vars='all', coords='different'):
 _CONCAT_DIM_DEFAULT = '__infer_concat_dim__'
 
 
+def _infer_concat_order_from_nested_list(datasets, concat_dims):
+
+    combined_ids = _infer_tile_ids_from_nested_list(datasets, [], {})
+
+    # Currently if concat_dims is not supplied then _auto_concat attempts to deduce it on every call
+    # TODO might be faster in this case to just work out the concat_dims once here
+    tile_id, ds = list(combined_ids.items())[0]
+    n_dims = len(tile_id)
+    if concat_dims is None or concat_dims == _CONCAT_DIM_DEFAULT:
+        concat_dims = [_CONCAT_DIM_DEFAULT]*n_dims
+    else:
+        if len(concat_dims) != n_dims:
+            raise ValueError("concat_dims has length " + str(len(concat_dims))
+                             + " but the datasets passed are nested in a " +
+                             str(n_dims) + "-dimensional structure")
+
+    return combined_ids, concat_dims
+
+
+def _infer_tile_ids_from_nested_list(entry, current_pos, combined_tile_ids):
+    """
+    Given a list of lists (of lists...) of datasets, returns a dictionary
+    with the index of each dataset in the nested list structure as the key.
+
+    Recursively traverses the given structure, while keeping track of the
+    current position.
+
+    Parameters
+    ----------
+    entry : list[list[xarray.Dataset, xarray.Dataset, ...]]
+        List of lists of arbitrary depth, containing datasets in the order they
+        are to be concatenated.
+
+    Returns
+    -------
+    combined_tile_ids : dict[tuple(int, ...), xarray.Dataset]
+    """
+
+    from .dataset import Dataset
+
+    if isinstance(entry, list):
+        # Dive down tree and recursively open the next list
+        current_pos.append(0)
+        for i, item in enumerate(entry):
+            current_pos[-1] = i
+            combined_tile_ids = _infer_tile_ids_from_nested_list\
+                (item, current_pos, combined_tile_ids)
+
+        # Move back up tree
+        del current_pos[-1]
+        return combined_tile_ids
+
+    elif isinstance(entry, Dataset):
+        # Termination condition
+        combined_tile_ids[tuple(current_pos)] = entry
+        return combined_tile_ids
+
+    else:
+        raise TypeError("Element at position " + str(tuple(current_pos)) +
+                        " is of type " + str(type(entry)) + ", which is "
+                        "neither a list nor an xarray.Dataset")
+
+
+def _check_shape_tile_ids(combined_tile_ids):
+    # TODO create custom exception class instead of using asserts?
```

Comment:
`ValueError` would be the usual error to raise here. And yes, this is probably a good idea -- otherwise somebody could try passing in a list with uneven depth, e.g., `[[ds1, ds2], ds3]`, which would probably result in a very difficult to debug error at some later point.

Reply:
Okay cool, I'll make sure that's covered sufficiently by the unit tests.